Lib/dataclasses.py

```diff
@@ -431,8 +431,8 @@ def _create_fn(name, args, body, *, globals=None, locals=None,
     # worries about external callers.
     if locals is None:
         locals = {}
-    if 'BUILTINS' not in locals:
```

Comment:
I'd suggest using `__BUILTINS__` because dunder names are supposed to be reserved for use by the stdlib.

Reply:
Before I commit this, I'd like to spend some time researching why this test is even present, instead of just unconditionally assigning to `locals`. At the very least it could use a comment.

Also, I'm not sure that exposing all of `builtins` in `locals` is a good idea, versus just exposing `builtins.object`.

Reply:
Good catch, looks like the relevant history is:
- https://github.com/python/cpython/pull/9518/files#diff-44ce2dc1c4922b2f5cf7631d8f86cc569a4c25eb003aaecdc2bc22eb9163d5f5L371-L375
- https://github.com/python/cpython/pull/8452

That is, I think this check was made dead in #9518, but wasn't noticed in that PR

Reply:
I can do some double checking and clean that up + change the PR to only pass along `object`.

(I'll also note that there's still the inscrutable `().__class__.__base__` option on the table, in case we don't want to expose anything at all)

Reply:
Okay, I audited all the call sites, it looks like there is one case where this check is not dead, but it's accidental.

Over here we reuse the same locals dict for two different `_create_fn` calls:
https://github.com/python/cpython/blob/87b5fd9d9de11c68b9940d3a1dba902138129109/Lib/dataclasses.py#L619

So my conclusion is:
a) It's safe to remove the check.
b) We should actually go a little further. Since we only need builtins for the frozen init, we should pass that in explicitly when creating `__init__` and remove this from `_create_fn`